allennlp/modules/seq2seq_encoders/efficient_multi_head_attention.py

```diff
@@ -0,0 +1,159 @@
+        init.xavier_normal(self._value_projections)
+
+    def get_input_dim(self):
+        return self._input_dim
+
+    def get_output_dim(self):
+        return self._output_dim
+
+    def forward(self,  # pylint: disable=arguments-differ
+                inputs: torch.Tensor,
+                mask: torch.LongTensor = None) -> torch.FloatTensor:
+        """
+        Parameters
+        ----------
+        inputs : ``torch.FloatTensor``, required.
+            A tensor of shape (batch_size, timesteps, input_dim)
+        mask : ``torch.FloatTensor``, optional (default = None).
+            A tensor of shape (batch_size, timesteps).
+
+        Returns
+        -------
+        A tensor of shape (batch_size, timesteps, output_projection_dim),
+        where output_projection_dim = input_dim by default.
+        """
+        num_heads = self._num_heads
+
+        batch_size, timesteps, hidden_dim = inputs.size()
+        if mask is None:
+            mask = Variable(inputs.data.new(batch_size, timesteps).fill_(1.0))
+
+        # Treat the queries, keys and values each as a ``num_heads`` size batch.
+        # shape (num_heads, batch_size * timesteps, hidden_dim)
+        inputs_per_head = inputs.repeat(num_heads, 1, 1).view(num_heads,
+                                                              batch_size * timesteps,
+                                                              hidden_dim)
+        # Do the projections for all the heads at once.
+        # Then reshape the result as though it had a
+        # (num_heads * batch_size) sized batch.
+        queries_per_head = torch.bmm(inputs_per_head, self._query_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        queries_per_head = queries_per_head.view(num_heads * batch_size,
+                                                 timesteps,
+                                                 self._attention_dim)
+
+        keys_per_head = torch.bmm(inputs_per_head, self._key_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        keys_per_head = keys_per_head.view(num_heads * batch_size,
+                                           timesteps,
+                                           self._attention_dim)
+
+        values_per_head = torch.bmm(inputs_per_head, self._value_projections)
+        # shape (num_heads * batch_size, timesteps, attention_dim)
+        values_per_head = values_per_head.view(num_heads * batch_size, timesteps, self._values_dim)
+
+        # shape (num_heads * batch_size, timesteps, timesteps)
+        scaled_similarities = torch.bmm(queries_per_head, keys_per_head.transpose(1, 2)) / self._scale
+
+        # shape (num_heads * batch_size, timesteps, timesteps)
+        # Normalise the distributions, using the same mask for all heads.
+        attention = last_dim_softmax(scaled_similarities, mask.repeat(num_heads, 1))
+        attention = self._attention_dropout(attention)
```

Comment:
I don't think I buy that this is still a well-defined probability distribution.  `attention.sum(dim=-1)` after line 138 should sum to 1, and `attention.sum(dim=-1)` will necessarily be less than one with non-zero dropout.  What am I missing?

Reply:
Yeah you're right, thanks.

Reply:
Also this seems a bit mad given this isn't the case? What does doing this even do?

Reply:
Does the transformer paper do this?  I just checked their paper and it wasn't clear to me whether they had this particular dropout here or not.  It doesn't seem particularly problematic - you're just saying that the model should be robust to having some of its dependent words missing.  It's just problematic to say that it's still a normalized distribution.